### PR TITLE
Implement a custom unittest runner + manual thread shutdown

### PR DESCRIPTION
**Pull request by AndrejMitrovic**
This should fix most issues with the random segfaults in ocean and when using SCP.

### CODE REVIEW

File: source/agora/test/Base.d [hunk 1]
```diff
@@ -45,8 +45,59 @@ import std.algorithm.iteration;
...
+            }
+            catch (Throwable ex)
+            {
+                writeln(ex);
+            }
+        }
+    }
+
+    //result.summarize = true;
+    result.runMain = false;
```
Comment (Geod24) [inline source/agora/test/Base.d:97]: <3

**Review by Geod24 — APPROVED** [review]
LVGTM

**Comment by AndrejMitrovic** [comment]
It seems one of the builds always stalls for a long while, and it's because of frequent failures to upload the coverage report to ~circlecli~ codecov:

Screen Shot 2019-10-14 at 10 23 58

**Comment by AndrejMitrovic** [comment]
Actually, how does coverage reporting work? We have multiple builds, so which coverage report is used for ~CircleCI~ codecov? Just whichever was the last build that finished?

Edit: I keep using the wrong word, sorry.

**Comment by Geod24** [comment]
> It seems one of the builds always stalls for a long while, and it's because of frequent failures to upload the coverage report to circlecli codecov:

Yeah I think this machine's network is bad

> Actually, how does coverage reporting work? We have multiple builds, so which coverage report is used for CircleCI codecov? Just whichever was the last build that finished?

It merges them

**Comment by AndrejMitrovic** [comment]
Great. And it's green.

**Pull request merged**